flux-refineck/src/constraint_gen.rs

```diff
@@ -236,13 +237,22 @@ impl<'a, 'tcx> ConstrGen<'a, 'tcx> {
             }
         }
 
+        // check (non-closure) obligations -- the closure ones are handled in `checker` since
+        // as we have to recursively walk over their def_id bodies.
+        for pred in &obligs {
+            if let rty::ClauseKind::Projection(projection_pred) = pred.kind().skip_binder() {
+                let impl_elem =
+                    normalize_alias_ty(infcx.genv, callsite_def_id, &projection_pred.alias_ty)?;
```

Comment:
Not sure about this, but instead of opening the predicate and then normalizing the `AliasTy` we could normalize the predicate (which is `TypeFoldable`) and when opened everything will be formalized.

Reply:
ah yes, I was trying to avoid exposing a second function, this is perfect, thanks!

Reply:
Hmm no, that doesn't quite work because we want to turn the `AliasTy` into a `Ty` but the "folder" for `AliasTy` returns an `AliasTy` ...

Reply:
I suppose we could turn the `alias_ty` into a proper `ty` and then just fold over that...

Reply:
oh right, we would need a more general predicate for equality between arbitrary types. ok then, so let's leave it like this.